### ISSUE TITLE
Infinite loop if something else sets errno

### ISSUE THREAD

**Issue by GreyHak** [ericcurtin/execxx#1]
When performing read I believe errno should only be checked if the return value is -1.  If something else in the program sets errno to EAGAIN or EINTR (in my case EINTR), and read were to return 0 (zero) on completion, the read loop ends up going infinitely.  I'm unsure if this is the best solution, but adding this worked for me...

```
    else if (r == 0) {
      errno = 0;  // and try again
    }
```

**Comment by ericcurtin** [comment]
Ah well spotted, maybe we should rewrite this in Rust sometime, looking for something not so difficult to rewrite to learn the language. Wanna collaborate? execrs maybe?

**Comment by ericcurtin** [comment]
Sorry was mixing this up with another similar project I have sxx. Do you have a simple code example to reproduce this? I can fix

**Comment by ericcurtin** [comment]
This project isn't really suited to a rust rewrite, it's just a header file

**Comment by GreyHak** [comment]
Thank you for the responsiveness @ericcurtin .  Here is a really simple test case...

`
[cent@localhost exetest]$ cat exetest.cpp
#include <iostream>
#include "execxx.hpp"

int main()
{
   (void)qx({"echo", "command1"});
   std::cerr << "Completed command 1" << std::endl;
   (void)qx({"echo", "command2"});
   std::cerr << "Completed command 2" << std::endl;
   errno = EINTR;
   (void)qx({"echo", "command3"});
   std::cerr << "Completed command 3" << std::endl;
   return 1;
}
[cent@localhost exetest]$ g++ exetest.cpp -o exetest
[cent@localhost exetest]$ ./exetest
Completed command 1
Completed command 2
^C
[cent@localhost exetest]$
`

**Comment by GreyHak** [comment]
Thank you for the responsiveness @ericcurtin .  Here is a really simple example which demonstrates the issue.  This was tested on CentOS 9 using g++ (GCC) 11.3.1.

[cent@localhost exetest]$ cat exetest.cpp
```
#include <iostream>
#include "execxx.hpp"

int main()
{
   (void)qx({"echo", "command1"});
   std::cerr << "Completed command 1" << std::endl;
   (void)qx({"echo", "command2"});
   std::cerr << "Completed command 2" << std::endl;
   errno = EINTR;
   (void)qx({"echo", "command3"});
   std::cerr << "Completed command 3" << std::endl;
   return 1;
}
```
[cent@localhost exetest]$ g++ exetest.cpp -o exetest
[cent@localhost exetest]$ ./exetest
Completed command 1
Completed command 2
^C
[cent@localhost exetest]$

**Comment by ericcurtin** [comment]
I'll push something tomorrow... Funnily enough I have a C version of this thing that I have been using recently that has fixed this bug I think, will check tomorrow

**Comment by ericcurtin** [comment]
Fixed in c21d9f13e11f828e8d46decefa49b11eb335ac33

Funnily enough I had this fixed in a C version of this code I use sometimes, just hadn't propagated it to here:

```
char* qx(const char* cmd) {
  int stdout_fds[2];
  if (pipe(stdout_fds) == -1) {
    PERR("pipe(%p)", stdout_fds);
    return NULL;
  }

  const pid_t pid = fork();
  if (!pid) {
    close(stdout_fds[0]);
    dup2(stdout_fds[1], 1);
    dup2(stdout_fds[1], 2);

    close(stdout_fds[1]);

    execl("/bin/sh", "sh", "-c", cmd, (char *) NULL);
    exit(0);
  }

  close(stdout_fds[1]);

  int buf_size = 4096;
  char* out = malloc(buf_size);
  ssize_t tot_r = 0;
  ssize_t r = 0;
  do {
    r = read(stdout_fds[0], out + tot_r, buf_size - tot_r - 1);
    if (r <= 0) {
      break;
    }

    tot_r += r;
    if (buf_size - tot_r - 1 <= 0) {
      buf_size *= 2;
      out = realloc(out, buf_size);
    }
  } while (r > 0 || !errno || errno == EAGAIN || errno == EINTR);

  out[tot_r] = 0;
  close(stdout_fds[0]);

  int ret, status;
  do {
    ret = waitpid(pid, &status, 0);
  } while (ret == -1 && errno == EINTR);

  return out;
}
```

Anyway, any more issues like this, feel free to open a PR
